### PR TITLE
Added blog and posts section

### PR DESCRIPTION
**Pull request by maceq687**
~~Created revops landing page based on Simon's design. Initial code comes from the Webflow project~~

- Added blog post page layout
- Created `/blog` site that lists all blog posts
- Added latest posts section to revops page
- Added latest posts section to services page layout
  - ~~specific services only list related blog posts (pipedrive, hubspot)~~

### CODE REVIEW

File: revops/revops.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    serviceItemIcon.src = service.Icon;
+    serviceItemIcon.classList.add("image");
+
+    const serviceItemTitle = document.createElement("h3");
+    serviceItemTitle.innerHTML = service.Name;
+    serviceItemTitle.classList.add("post-preview-title");
+
+    const serviceItemSummary = document.createElement("p");
+    serviceItemSummary.innerHTML = service["Short description"];
+    serviceItemSummary.classList.add("post-summary");
```
Comment (hypesystem) [inline revops/revops.js]: Instead of building elements like this, I usually create a top-level element and create the rest of the structure with `innerHTML`:

```js
const servicesList = document.createElement("div");
servicesList.classList.add("collection-list", "w-dyn-items");
servicesList.innerHTML = services.map((service) => `
  <div class="collection-item w-dyn-item">
    <a href="${service.Slug}" class="link-block w-inline-block">
      <div class="div-block">
        <img src="${service.Icon}">
        <h3 class="post-preview-title">${service.Name}</h3>
        <p class="post-summary">${service["Short description"]}</p>
      </div>
    </a>
  </div>
`).join("");
```

This leans into the structure of HTML and keeps us closer to a single way of defining HTML.

File: revops/blog-posts.json [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    "Post Summary": "Cras sit amet velit id nulla tempus dictum sit amet eu nisi.\nMaecenas euismod sapien eu arcu convallis, vitae vestibulum ipsum maximus.\nPraesent nec orci at nulla consequat congue ut non arcu.\nLorem ipsum dolor sit amet, consectetur adipiscing elit.\nPraesent nec orci at nulla consequat congu",
+    "Main Image": "https://uploads-ssl.webflow.com/66c31735a489578abdbf26f1/66c3422c3188060700245da4_image18.jpeg",
+    "Thumbnail image": "https://uploads-ssl.webflow.com/66c31735a489578abdbf26f1/66c31798be4b2f0a05bc84c7_image16.jpeg",
+    "Color": "#017737",
+    "Tags": "pipedrive",
+    "Date": "Sun Aug 11 2024 17:13:14 GMT+0000 (Coordinated Universal Time)",
+    "CRM service": "pipedrive"
+  }
+]
```
Comment (hypesystem) [inline revops/blog-posts.json]: I love the idea of bootstrapping a CMS by exporting json files from somewhere Simon can edit them - great idea

File: revops/revops.js [hunk 3]
```diff
@@ -45,4 +47,38 @@ async function populateServices() {
...
+    const blogPostThumbnail = clone.getElementById("post-preview-thumbnail");
+    blogPostThumbnail.src = post["Thumbnail image"];
+
+    const blogPostTitle = clone.getElementById("post-preview-title");
+    blogPostTitle.innerHTML = post.Name;
+
+    const blogPostSummary = clone.getElementById("post-preview-summary");
+    blogPostSummary.innerHTML = post["Post Summary"];
+
+    blogPostsList.appendChild(clone);
```
Comment (hypesystem) [inline revops/revops.js]: C.f. other note on how to generate HTML in native Javascript.

But this makes me thing something else: instead of fetching the data in the browser when the page loads, you can just generate a page per blog post. C.f. how it is done in Jekyll: https://jekyllrb.com/docs/collections/

**Review by hypesystem — CHANGES_REQUESTED** [review]
Overall a lot of this makes sense.

I think you should lean more into the static-site generation of Jekyll, though, c.f. https://jekyllrb.com/docs/collections/

**Comment by hypesystem** [comment]
I fixed the build by updating Ruby and Jekyll versions we use. See it here: https://deploy-preview-93--zealous-fermat-77bc7b.netlify.app/revops/

Visual feedback:

- We try to stay consistent by using the same menu (burger menu opens menu items) and logo (text top-left) with none of these two elements inside a specific header. See most of the other pages for this pattern. It's one of the few constraints we have landed on for the design.
- We already have a button design. Reuse it to be consistent.
- We have a consistent footer across pages. Reuse it.

Other than that it looks interesting, and I look forward to seeing where it goes :-)

Make sure we are in the clear with copyright on e.g. images used

File: revops/revops.js (hunk 1 above)
Comment (maceq687) [reply revops/revops.js]: Yes, I agree building HTML the way I did it here is inefficient, I have used different method for the 'latest blog posts' list, utilizing `<template>`, but what you are suggesting might be even better

File: revops/blog-posts.json (hunk 2 above)
Comment (maceq687) [reply revops/blog-posts.json]: Yes, once we will figure out what field are needed for the json file we could potentially make it work this way.

**Comment by hypesystem** [comment]
Very cool changes!

I think for the RevOps page it might make sense to make the hello section a bit less tall. The frontpage is very "vibes" focused - but if people have already segmented into revops, we can get to the point a bit quicker. But check with Simon if he agrees that this makes sense

File: revops/revops.js [hunk 4]
```diff
@@ -51,16 +51,26 @@ async function populateBlogPosts() {
   const request = new Request("blog-posts.json");
 
   const response = await fetch(request);
-  blogPosts = await response.json();
+  var unorderedBlogPosts = await response.json();
+
+  blogPosts = unorderedBlogPosts.sort(function (a, b) {
+    if (new Date(a.Date) == new Date(b.Date)) return 0;
+    if (new Date(a.Date) < new Date(b.Date)) return -1;
+    if (new Date(a.Date) > new Date(b.Date)) return 1;
```
Comment (hypesystem) [inline revops/revops.js]: Code style:

I like the if-return style here. In order to keep the code readable (and actually more performant!) I would suggest reducing the complexity by extracting variables.

I would also encourage being consistent with indentation levels and indications of nested blocks. That is, *always* use curly braces when specifying a block, and *always* indent the same level. This makes it easier to quickly parse code, as the indentation levels give a quick indication of what is related and how.

```js
const aDate = new Date(a.Date);
const bDate = new Date(b.Date);
if (aDate == bDate) {
  return 0;
}
if(aDate < bDate) {
  return -1;
}
if(aDate > bDate) {
  return 1;
}
```

This is *more lines* but that's not a problem. Lines are free.

On the other hand it is more *honest* about the complexity of the function, instead of trying to hide the complexity by fitting as much as possible into each line.

It's like the difference between writing long, convoluted paragraph-length sentences to seem clever, or breaking writing into shorter, more easily digestible sentences. When the goal is only clarity of communication, it is *always* better to err on the side of shorter and clearer sentences.

File: revops/revops.js (hunk 1 above)
Comment (maceq687) [reply revops/revops.js]: Fixed

File: revops/revops.js (hunk 4 above)
Comment (maceq687) [reply revops/revops.js]: Yes, makes total sense, fixed

File: revops/revops.js (hunk 3 above)
Comment (maceq687) [reply revops/revops.js]: I have now rebuilt the services and blog posts structures using Jekyll collections

File: revops/services.json [hunk 5]
```diff
@@ -53,7 +53,7 @@
     "Created On": "Mon Aug 19 2024 09:59:52 GMT+0000 (Coordinated Universal Time)",
     "Updated On": "Mon Aug 19 2024 10:04:03 GMT+0000 (Coordinated Universal Time)",
     "Published On": "",
-    "Icon": "https://uploads-ssl.webflow.com/66c31735a489578abdbf26f1/66c31798be4b2f0a05bc84d2_image11.jpeg",
+    "Icon": "../img/Pipedrive Premier Partner Denmark deranged.png",
```
Comment (hypesystem) [inline revops/services.json]: Avoid relative paths for things like this. Instead, `/img/....png` would work.

We don't know if the content from services.json will always be rendered on the same page, and the image references should always work!

File: revops/services.json [hunk 6]
```diff
@@ -38,7 +38,7 @@
     "Created On": "Mon Aug 19 2024 09:59:52 GMT+0000 (Coordinated Universal Time)",
     "Updated On": "Mon Aug 19 2024 10:08:58 GMT+0000 (Coordinated Universal Time)",
     "Published On": "",
-    "Icon": "https://uploads-ssl.webflow.com/66c31735a489578abdbf26f1/66c31798be4b2f0a05bc84c7_image16.jpeg",
```
Comment (hypesystem) [inline revops/services.json]: Can we get these exported as SVGs? Or PNGs with transparent background on what shouldn't be filled? I think that would look nicer

File: revops/index.html [hunk 7]
```diff
@@ -64,7 +64,7 @@ <h2 class="section-heading">Our Expertise</h2>
 <section class="section">
   <div class="container-home-page w-container">
     <div class="section-title-group">
-      <h4>Entrusted by:</h4>
+      <h4>deranged is entrusted by:</h4>
```
Comment (hypesystem) [inline revops/index.html]: "entrusted" here feels off to me, maybe rephrase? Maybe just "trusted"?

File: _posts/2022-12-04-peanut-butter-jelly.md [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+</section>
```
Comment (hypesystem) [inline _posts/2022-12-04-peanut-butter-jelly.md]: If you nest things in HTML, don't use `.md` format - it skips rendering anyway.

Use markdown files for markdown content and HTML files for HTML content.

It's fine, imo, to use HTML for the blog posts.

File: css/revops.css [hunk 9]
```diff
@@ -30,7 +30,7 @@ h3 {
   margin-bottom: 0;
   font-size: 20px;
   /* font-weight: 300; */
-  line-height: 30px;
+  /* line-height: 30px; */
```
Comment (hypesystem) [inline css/revops.css]: Why have commented out CSS? Remove it - we can always find it in version control later if you really need it.

File: _layouts/services.html [hunk 10]
```diff
@@ -49,7 +49,7 @@
 </style>
 <section class="header">
   <a href="/" class="logo">deranged</a>
-  <h1><a href="/revops">RevOps services</a></h1>
+  <h1><a href="/revops#expertise">services</a></h1>
```
Comment (hypesystem) [inline _layouts/services.html]: I feel like these may be revops specific services? If so, that would make sense to clarify.

**Review by hypesystem — COMMENTED** [review]
I feel like this approach with placeholder content may move you further away from being able to release quickly when you get something to put out.

This PR is also very big and messy. I would prefer smaller PRs with little changes now that we have something out.

Your next goal should be to get rid of this big PR - by merging or splitting - so we have smaller things we can look at individually.

Some of the structural changes could easily be merged and released (e.g. services and blog posts as collections, but leaving the collection empty so no pages are generated) but because it is all in a big bunch here it's harder to get something out.

File: _config.yml [hunk 11]
```diff
@@ -1,7 +1,13 @@
+plugins:
+  - jekyll-paginate
```
Comment (hypesystem) [inline _config.yml]: Getting this error when building now:

```
run into trouble, you can find helpful resources at https://jekyllrb.com/help/! 
5:27:05 PM: jekyll 3.9.4 | Error:  jekyll-paginate
5:27:05 PM: /opt/buildhome/.rvm/gems/ruby-3.3.2/gems/jekyll-3.9.4/lib/jekyll/external.rb:68:in `rescue in block in require_with_graceful_fail': jekyll-paginate (Jekyll::Errors::MissingDependencyException)
```

I have never used a Jekyll plugin before and am not clear on how to approach it. How difficult would it be to do pagination without the plugin?

File: css/revops.css [hunk 12]
```diff
@@ -259,6 +259,10 @@ p {
   display: grid;
 }
 
+.blog-button {
```
Comment (hypesystem) [inline css/revops.css]: This could probably be made more generic, so it can be reused. Describe what it does instead of where it is used (that's actually a good rule of thumb for components in CSS, whereas page-specific styling can be named for where it appears)

File: revops/services.json (hunk 5 above)
Comment (maceq687) [reply revops/services.json]: Agreed, fixed

File: revops/index.html (hunk 7 above)
Comment (maceq687) [reply revops/index.html]: Changed to "trusted"

File: _posts/2022-12-04-peanut-butter-jelly.md (hunk 8 above)
Comment (maceq687) [reply _posts/2022-12-04-peanut-butter-jelly.md]: Agreed, I have later fixed my mistake in 03d441eaef7ab0bd9ebe1a8d4d18465489f814cd

File: _layouts/services.html (hunk 10 above)
Comment (maceq687) [reply _layouts/services.html]: Agreed, I'm still not 100% sure regarding navigation, it might need a revisit (including blog page and post layout as well)

File: css/revops.css (hunk 12 above)
Comment (maceq687) [reply css/revops.css]: Agreed, fixed

File: css/revops.css (hunk 9 above)
Comment (maceq687) [reply css/revops.css]: I was suppose to revisit the file to clear those and find if any more values are unused, I've cleaned it up now

File: revops/services.json (hunk 6 above)
Comment (maceq687) [reply revops/services.json]: I will ask Simon once he will be back

File: _config.yml (hunk 11 above)
Comment (maceq687) [reply _config.yml]: I can write it up in JS, but I think that making use of the built-in solution might make sense here, we probably won't need it to do anything fancier than the bare bones pagination with control that it offers.
It should be possible to install it on the docker image like this:
```
 command: gem install jekyll-paginate && jekyll serve --watch --force_polling --verbose --livereload
 ```

Comment (maceq687) [reply _config.yml]: Or perhaps we could use "infinite" scrolling to load more posts with lazy method instead of pagination? https://jekyllcodex.org/without-plugin/lazy-loading/

**Comment by maceq687** [comment]
I've moved some of the changes that were made here to a separate branch that could be merged first (https://github.com/derangeddk/deranged.dk/pull/97)
This branch should bring less changes after the other one is merged and this one is rebased

File: _config.yml [hunk 13]
```diff
@@ -0,0 +1,7 @@
+collections:
+  services:
+    output: true
+    permalink: revops/:collection/:name
+  posts:
+    output: true
+    permalink: revops/:collection/:name
```
Comment (hypesystem) [inline _config.yml]: I think it might make sense i blog posts just live at `/:blog/:name` so we can have a single blog for everything. Then we can tag or categorize blog posts to make them appear on particular pages, e.g. /revops could get blog posts filtered by having the revops tag and show a few of them on the page.

File: _config.yml [hunk 14]
```diff
@@ -0,0 +1,7 @@
+collections:
+  services:
+    output: true
+    permalink: revops/:collection/:name
```
Comment (hypesystem) [inline _config.yml]: Are services a thing under revops? Or is it a general thing you want to introduce?

If e.g. pipedrive or fractional COO are services, I think they should have top-level pages, e.g. deranged.dk/pipedrive and deranged.dk/fractional-coo or just deranged.dk/coo so they are easy to share.

Placing them under /services doesn't add anything imo?

File: _layouts/post.html [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+<section class="section">
+  <div class="container-home-page w-container">
+    <div class="section-title-group">
+      <h2 class="section-heading">Recent posts</h2>
+    </div>
+    <div class="latest-posts w-dyn-list">
+      <div class="latest-posts-list w-dyn-items">
+        {% assign filtered_posts = site.posts | where_exp: "post", "post.title
+        != page.title" %} {% for post in filtered_posts limit:3 %}
```
Comment (hypesystem) [inline _layouts/post.html]: This is impossible to read :smile:

File: _layouts/post.html [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+              />
+              <h3 class="post-preview-title">{{post.title}}</h3>
+              <p class="post-summary">{{post.excerpt}} <u>Read&nbsp;more</u></p>
+            </div>
+          </a>
+        </div>
+        {% endfor %}
+      </div>
+    </div>
+    <a href="/revops/blog" class="button width-auto">Read more posts here</a>
```
Comment (hypesystem) [inline _layouts/post.html]: I think it would make sense to think of what else we could offer visitors.

We could link to:

- cases page if they want to read more about our work
- the team page if they want more about us
- explainer page for in-depth explanations of industry relevant topics

File: _layouts/services.html [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+              <p class="post-summary">{{post.excerpt}} <u>Read&nbsp;more</u></p>
+            </div>
+          </a>
+        </div>
+        {% endfor %}
+      </div>
+    </div>
+    <a href="/revops/blog" class="button width-auto">Read more posts here</a>
+  </div>
+</section>
```
Comment (hypesystem) [inline _layouts/services.html]: I'm a bit confused about what a service is in this conception :smile: or why it needs both a collection and a layout that differs from what already exists.

Right now it seems very similar to blog post pages, which is also odd if it is just about things we do.

File: _services/fractional-coo.html [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+  <p>
+    Integer rutrum ante et nunc venenatis, id ultricies risus ultricies.
+    Maecenas euismod sapien eu arcu convallis, vitae vestibulum ipsum maximus.
+    Donec eu felis at libero consequat sagittis a et urna. Praesent nec orci at
+    nulla consequat congue ut non arcu. Praesent nec orci at. Bridge the gap
+    between appointments or reduce your costs by hiring a fractional executive.
+    We bring our expertise with us whether you’re looking to make process
+    improvements, manage specific projects or oversee daily operations.
+  </p>
+</section>
```
Comment (hypesystem) [inline _services/fractional-coo.html]: How long do you envision services pages being? What content do you think we should have on there?

We have a mindmap of different "services" or similar we could add to the TODO pile, and I'm happy to write some up... but I think we need to talk about how we pitch a service well :smile:

Comment (hypesystem) [reply _services/fractional-coo.html]: I can try and write a pitch of Business Central for a page if you would like it? Then we can talk about my considerations and compare/contrast with Simon's takes, to land on some general principles for what goes on these pages?

File: revops/index.html [hunk 19]
```diff
@@ -56,15 +61,30 @@ <h2 class="section-heading">Our Expertise</h2>
       </div>
     </div>
     <div class="collection-list-wrapper w-dyn-list">
-      <div class="collection-list w-dyn-items" id="services"></div>
+      <div class="collection-list w-dyn-items">
+        {% for s in page.services %} {% for service in site.services %} {% if s
+        == service.slug %}
```
Comment (hypesystem) [inline revops/index.html]: Cannae read this

File: revops/index.html [hunk 20]
```diff
@@ -56,15 +61,30 @@ <h2 class="section-heading">Our Expertise</h2>
...
+        <div class="collection-item w-dyn-item">
+          <a href="{{service.url}}" class="link-block w-inline-block">
```
Comment (hypesystem) [inline revops/index.html]: Until we have more content on the services we should remove the link imo.

We can launch with just the excerpts here, but without showing the main pages.

**Review by hypesystem — COMMENTED** [review]
Looks good. I have some thoughts on various things.

As before, I think the goal should be finding the slice you can make to get this out quickly.

That may mean:

- if we have no blog posts, hide the relevant sections on the revops page
- remove links to services pages and write quick texts for the service excerpts so we can get something out

That way you can get the structural work out and start on new PRs to keep it easily reviewable.

**Comment by hypesystem** [comment]
I'm confused about the role of this PR now as you have pulled out other PRs?

File: revops/services.json (hunk 6 above)
Comment (maceq687) [reply revops/services.json]: See: https://github.com/derangeddk/deranged.dk/pull/99

File: _config.yml (hunk 13 above)
Comment (maceq687) [reply _config.yml]: We (me and Simon) have originally assumed that the blog and posts will be part of revops only, if there is plan to make it cover a broader spectrum of topics then of course it makes sense to redo the navigation and it's logic and make it work on it's own and just categorize revops related posts accordingly

File: revops/index.html (hunk 20 above)
Comment (maceq687) [reply revops/index.html]: Already addressed in: https://github.com/derangeddk/deranged.dk/pull/97

File: _config.yml (hunk 14 above)
Comment (maceq687) [reply _config.yml]: Addressed in https://github.com/derangeddk/deranged.dk/commit/74a3976d56193c03078f443fd2f16c56f800b713

File: _config.yml (hunk 13 above)
Comment (maceq687) [reply _config.yml]: Addressed here: https://github.com/derangeddk/deranged.dk/pull/100/commits/4d87e4681159b5ce301e19d0401ee181a0b6fb93

File: _layouts/post.html (hunk 15 above)
Comment (maceq687) [reply _layouts/post.html]: Addressed in https://github.com/derangeddk/deranged.dk/pull/93/commits/1e5c0604e0d2d9d0006ec56983b1daf5f35d37aa

File: _layouts/services.html (hunk 17 above)
Comment (maceq687) [reply _layouts/services.html]: See content added to revops services in PR https://github.com/derangeddk/deranged.dk/pull/100

File: _services/fractional-coo.html (hunk 18 above)
Comment (maceq687) [reply _services/fractional-coo.html]: See content added to revops services in PR https://github.com/derangeddk/deranged.dk/pull/100

File: revops/index.html (hunk 19 above)
Comment (maceq687) [reply revops/index.html]: Addressed in https://github.com/derangeddk/deranged.dk/pull/100/commits/4c2a1982ecdbe8aa65d3400b8e8b9fd0be703bd8

**Comment by maceq687** [comment]
> I'm confused about the role of this PR now as you have pulled out other PRs?

Right now this PR bring changes related to adding posts collection and the `/blog` page

File: _services/fractional-coo.html (hunk 18 above)
Comment (maceq687) [reply _services/fractional-coo.html]: Content related to DevOps and other services offered by deranged is more than welcome 🙂

**Comment by hypesystem** [comment]
@maceq687 Can we rename, rebase the PR so the purpose is clear?

**Comment by hypesystem** [comment]
@maceq687 Can we close this PR now? Or what's left in here that should make it into smaller PRs?
